tests/vanilla/shallow.test.tsx

```diff
@@ -172,6 +172,25 @@ describe('shallow', () => {
   })
 })
 
+describe('mixed cases', () => {
+  const obj = { 0: 'foo', 1: 'bar' }
+  const arr = ['foo', 'bar']
+  const set = new Set(['foo', 'bar'])
+  const map = new Map([
+    [0, 'foo'],
+    [1, 'bar'],
+  ])
+
+  it('compares different data structures', () => {
+    expect(shallow<unknown>(obj, arr)).toBe(false)
+    expect(shallow<unknown>(obj, set)).toBe(false)
+    expect(shallow<unknown>(obj, map)).toBe(false)
```

Comment:
```js
  const obj = { 0: 'foo', 1: 'bar' }
  const map = new Map([[0, 'foo'], [1, 'bar']])
```
Are they shallow equal, or not? (It feels like our current implementation doesn't handle it very well.)

Reply:
https://github.com/pmndrs/zustand/pull/3108#discussion_r2075038796